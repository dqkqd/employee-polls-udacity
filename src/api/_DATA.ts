import type {
  AnswerId,
  Question,
  QuestionId,
  QuestionsDictionary,
  User,
  UserId,
  UsersDictionary,
} from "../interfaces"

let users: UsersDictionary = {
  sarahedo: {
    id: "sarahedo",
    password: "password123",
    name: "Sarah Edo",
    avatarURL: null,
    answers: {
      "8xf0y6ziyjabvozdd253nd": "optionOne",
      "6ni6ok3ym7mf1p33lnez": "optionOne",
      am8ehyc8byjqgar0jgpub9: "optionTwo",
      loxhs1bqm25b708cmbf3g: "optionTwo",
    },
    questions: ["8xf0y6ziyjabvozdd253nd", "am8ehyc8byjqgar0jgpub9"],
  },
  tylermcginnis: {
    id: "tylermcginnis",
    password: "abc321",
    name: "Tyler McGinnis",
    avatarURL: null,
    answers: {
      vthrdm985a262al8qx3do: "optionOne",
      xj352vofupe1dqz9emx13r: "optionTwo",
    },
    questions: ["loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do"],
  },
  mtsamis: {
    id: "mtsamis",
    password: "xyz123",
    name: "Mike Tsamis",
    avatarURL: null,
    answers: {
      xj352vofupe1dqz9emx13r: "optionOne",
      vthrdm985a262al8qx3do: "optionTwo",
      "6ni6ok3ym7mf1p33lnez": "optionOne",
    },
    questions: ["6ni6ok3ym7mf1p33lnez", "xj352vofupe1dqz9emx13r"],
  },
  zoshikanlu: {
    id: "zoshikanlu",
    password: "pass246",
    name: "Zenobia Oshikanlu",
    avatarURL: null,
    answers: {
      xj352vofupe1dqz9emx13r: "optionOne",
    },
    questions: [],
  },
}

let questions: QuestionsDictionary = {
  "8xf0y6ziyjabvozdd253nd": {
    id: "8xf0y6ziyjabvozdd253nd",
    author: "sarahedo",
    timestamp: 1467166872634,
    optionOne: {
      votes: ["sarahedo"],
      text: "Build our new application with Javascript",
    },
    optionTwo: {
      votes: [],
      text: "Build our new application with Typescript",
    },
  },
  "6ni6ok3ym7mf1p33lnez": {
    id: "6ni6ok3ym7mf1p33lnez",
    author: "mtsamis",
    timestamp: 1468479767190,
    optionOne: {
      votes: [],
      text: "hire more frontend developers",
    },
    optionTwo: {
      votes: ["mtsamis", "sarahedo"],
      text: "hire more backend developers",
    },
  },
  am8ehyc8byjqgar0jgpub9: {
    id: "am8ehyc8byjqgar0jgpub9",
    author: "sarahedo",
    timestamp: 1488579767190,
    optionOne: {
      votes: [],
      text: "conduct a release retrospective 1 week after a release",
    },
    optionTwo: {
      votes: ["sarahedo"],
      text: "conduct release retrospectives quarterly",
    },
  },
  loxhs1bqm25b708cmbf3g: {
    id: "loxhs1bqm25b708cmbf3g",
    author: "tylermcginnis",
    timestamp: 1482579767190,
    optionOne: {
      votes: [],
      text: "have code reviews conducted by peers",
    },
    optionTwo: {
      votes: ["sarahedo"],
      text: "have code reviews conducted by managers",
    },
  },
  vthrdm985a262al8qx3do: {
    id: "vthrdm985a262al8qx3do",
    author: "tylermcginnis",
    timestamp: 1489579767190,
    optionOne: {
      votes: ["tylermcginnis"],
      text: "take a course on ReactJS",
    },
    optionTwo: {
      votes: ["mtsamis"],
      text: "take a course on unit testing with Jest",
    },
  },
  xj352vofupe1dqz9emx13r: {
    id: "xj352vofupe1dqz9emx13r",
    author: "mtsamis",
    timestamp: 1493579767190,
    optionOne: {
      votes: ["mtsamis", "zoshikanlu"],
      text: "deploy to production once every two weeks",
    },
    optionTwo: {
      votes: ["tylermcginnis"],
      text: "deploy to production once every month",
    },
  },
}

function generateUID() {
  return (
    Math.random().toString(36).substring(2, 15) +
    Math.random().toString(36).substring(2, 15)
  )
}

export function _getUsers() {
  return new Promise<UsersDictionary>((resolve) => {
    setTimeout(() => resolve({ ...users }), 1000)
  })
}

export function _saveUser(user: {
  id: string
  password: string
  name: string
  avatarURL?: string
}) {
  return new Promise<User>((resolve, reject) => {
    const { id, password, name, avatarURL } = user
    if (!id || !name || !password) {
      reject("Please provide id, name, password")
    } else if (Object.hasOwn(users, id)) {
      reject("User already existed")
    } else {
      const formattedUser = {
        ...user,
        avatarURL: avatarURL ?? null,
        answers: {},
        questions: [],
      }

      setTimeout(() => {
        users = {
          ...users,
          [id]: formattedUser,
        }

        resolve(formattedUser)
      }, 50)
    }
  })
}

export function _getQuestions() {
  return new Promise<QuestionsDictionary>((resolve) => {
    setTimeout(() => resolve({ ...questions }), 1000)
  })
}

function formatQuestion(question: {
  optionOneText: string
  optionTwoText: string
  author: UserId
}): Question {
  return {
    id: generateUID(),
    author: question.author,
    timestamp: Date.now(),
    optionOne: {
      votes: [],
      text: question.optionOneText,
    },
    optionTwo: {
      votes: [],
      text: question.optionTwoText,
    },
  }
}

export function _saveQuestion(question: {
  optionOneText?: string | null
  optionTwoText?: string | null
  author?: UserId | null
}) {
  return new Promise<Question>((resolve, reject) => {
    const { optionOneText, optionTwoText, author } = question
    if (!optionOneText || !optionTwoText || !author) {
      reject("Please provide optionOneText, optionTwoText, and author")
    } else if (optionOneText === optionTwoText) {
      reject("Options must be different")
    } else if (!Object.hasOwn(users, author)) {
      reject(`User id '${author}' does not exist`)
    } else {
      const formattedQuestion = formatQuestion({
        optionOneText,
        optionTwoText,
        author,
      })
      setTimeout(() => {
        questions = {
          ...questions,
          [formattedQuestion.id]: formattedQuestion,
        }

        resolve(formattedQuestion)
      }, 1000)
    }
  })
}

export function _saveQuestionAnswer(answer: {
  authedUser?: UserId | null
  qid?: QuestionId | null
  answerId?: AnswerId | null
}) {
  return new Promise<boolean>((resolve, reject) => {
    const { authedUser, qid, answerId } = answer

    if (!authedUser || !qid || !answerId) {
      reject("Please provide authedUser, qid, and answer")
    } else if (!Object.hasOwn(users, authedUser)) {
      reject(`User id '${authedUser}' does not exist`)
    } else if (!Object.hasOwn(questions, qid)) {
      reject(`Question id '${qid}' does not exist`)
    } else if (!["optionOne", "optionTwo"].includes(answerId)) {
      reject(`Answer should be 'optionOne' or 'optionTwo', not '${answerId}'`)
    } else {
      setTimeout(() => {
        users = {
          ...users,
          [authedUser]: {
            ...users[authedUser],
            answers: {
              ...users[authedUser].answers,
              [qid]: answerId,
            },
          },
        }

        questions = {
          ...questions,
          [qid]: {
            ...questions[qid],
            [answerId]: {
              ...questions[qid][answerId],
              votes: questions[qid][answerId].votes.concat([authedUser]),
            },
          },
        }

        resolve(true)
      }, 500)
    }
  })
}
