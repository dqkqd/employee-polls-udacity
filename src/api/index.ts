import {
  _getQuestions,
  _getUser,
  _getUsers,
  _saveQuestion,
  _saveQuestionAnswer,
  _saveUser,
} from "./_DATA"

export {
  _getQuestions as getQuestions,
  _getUser as getUser,
  _getUsers as getUsers,
  _saveQuestion as saveQuestion,
  _saveQuestionAnswer as saveQuestionAnswer,
  _saveUser as saveUser,
}
