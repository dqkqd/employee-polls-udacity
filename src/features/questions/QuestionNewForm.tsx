import {
  Button,
  CircularProgress,
  FormControl,
  TextField,
  Typography,
} from "@mui/material"
import { useState } from "react"
import { useNavigate } from "react-router-dom"
import { useAppDispatch, useAuth } from "../../app/hook"
import { createNewQuestion } from "../common"

const QuestionNewForm = () => {
  const dispatch = useAppDispatch()
  const navigate = useNavigate()

  const auth = useAuth()

  const [optionOneText, setOptionOneText] = useState("")
  const [optionTwoText, setOptionTwoText] = useState("")

  const [addStatus, setAddStatus] = useState<
    "idle" | "loading" | "succeeded" | "failed"
  >("idle")

  const loading = addStatus === "loading"
  const buttonEnabled = Boolean(
    !loading &&
      optionOneText &&
      optionTwoText &&
      optionOneText !== optionTwoText,
  )

  const sameOptionsErrorMassage =
    optionOneText && optionTwoText && optionOneText === optionTwoText
      ? "Two options should be different"
      : ""

  const [submitErrorMessage, setSubmitErrorMessage] = useState("")

  const handleAddQuestion = async (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault()
    setAddStatus("loading")
    const response = await dispatch(
      createNewQuestion({
        optionOneText,
        optionTwoText,
        author: auth.id as string,
      }),
    ).unwrap()
    if (response.ok && response.qid) {
      navigate(`/questions/${response.qid}`)
    } else {
      setSubmitErrorMessage(response.error ?? "")
    }
  }

  return (
    <>
      <FormControl>
        <TextField
          value={optionOneText}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setOptionOneText(e.target.value.trim())
          }
          label="First option"
          disabled={loading}
        />
      </FormControl>

      <FormControl>
        <TextField
          value={optionTwoText}
          onChange={(e: React.ChangeEvent<HTMLInputElement>) =>
            setOptionTwoText(e.target.value.trim())
          }
          label="Second option"
          disabled={loading}
        />
      </FormControl>

      <Button
        type="submit"
        variant="contained"
        onClick={handleAddQuestion}
        disabled={!buttonEnabled}
      >
        {loading ? (
          <CircularProgress data-testid="add-question-loading" />
        ) : (
          "Submit"
        )}
      </Button>

      {(sameOptionsErrorMassage || submitErrorMessage) && (
        <Typography fontSize={13} color="red">
          {sameOptionsErrorMassage || submitErrorMessage}
        </Typography>
      )}
    </>
  )
}

export default QuestionNewForm
