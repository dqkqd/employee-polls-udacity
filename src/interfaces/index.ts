export type UserId = string;
export type QuestionId = string;
export type AnswerId = "optionOne" | "optionTwo";

export interface Question {
  id: QuestionId;
  author: UserId;
  timestamp: number;

  optionOne: {
    votes: UserId[];
    text: string;
  };

  optionTwo: {
    votes: UserId[];
    text: string;
  };
}

export interface QuestionsDictionary {
  [id: QuestionId]: Question;
}

export interface User {
  id: UserId;
  name: string;
  password: string;
  avatarURL: null;
  questions: QuestionId[];
  answers: {
    [id: QuestionId]: AnswerId;
  };
}

export interface UsersDictionary {
  [id: UserId]: User;
}
