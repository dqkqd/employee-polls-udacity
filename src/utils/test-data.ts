import { EntityState } from "@reduxjs/toolkit"
import {
  getQuestions,
  getUsers,
  saveQuestion,
  saveQuestionAnswer,
  saveUser,
} from "../api"
import { AuthedUser, Question, User } from "../interfaces"
import { AnswerId } from "./../interfaces/index"

const fakeUsers = [
  {
    id: "@fake-user-1",
    name: "Fake User 1",
    password: "@fake-user-1-password",
    avatarURL: "https://picsum.photos/200",
  },
  {
    id: "@fake-user-2",
    name: "Fake User 2",
    password: "@fake-user-2-password",
    avatarURL: "https://picsum.photos/200",
  },
]

const fakeQuestions = [
  {
    optionOneText: "fake-question-1-option-one",
    optionTwoText: "fake-question-1-option-two",
    author: fakeUsers[0].id,
  },
  {
    optionOneText: "fake-question-2-option-one",
    optionTwoText: "fake-question-2-option-two",
    author: fakeUsers[1].id,
  },
  {
    optionOneText: "fake-question-3-option-one",
    optionTwoText: "fake-question-3-option-two",
    author: fakeUsers[1].id,
  },
]

const fakeQuestionsAnswers = [
  {
    userIndex: 0,
    questionIndex: 2, // fake-question-3
    answerId: "optionOne",
  },
]

export const initialAuth: AuthedUser = { ...fakeUsers[0], status: "success" }

const setupTestData = async () => {
  const addedUsers = await Promise.all(
    fakeUsers.map(async (user) => await saveUser(user)),
  )

  const addedQuestions = await Promise.all(
    fakeQuestions.map(async (question) => await saveQuestion(question)),
  )

  const addedAnswer = await Promise.all(
    fakeQuestionsAnswers.map(
      async ({ userIndex, questionIndex, answerId }) =>
        await saveQuestionAnswer({
          authedUser: addedUsers[userIndex].id,
          qid: addedQuestions[questionIndex].id,
          answerId: answerId as AnswerId,
        }),
    ),
  )

  if (!addedAnswer.every(Boolean)) {
    throw new Error("Some questions answer are not added")
  }

  const [allUsers, allQuestions] = await Promise.all([
    getUsers(),
    getQuestions(),
  ])
  const initialUsers: EntityState<User> = {
    ids: Object.keys(allUsers),
    entities: allUsers,
  }
  const initialQuestions: EntityState<Question> = {
    ids: Object.keys(allUsers),
    entities: allQuestions,
  }

  // need to fetch added users and added questions again because we added answer
  const addedUserIds = addedUsers.map((user) => user.id)
  const fetchAddedUsers = Object.values(allUsers).filter((user) =>
    addedUserIds.includes(user.id),
  )
  const addedQuestionIds = addedQuestions.map((question) => question.id)
  const fetchAddedQuestions = Object.values(allQuestions).filter((question) =>
    addedQuestionIds.includes(question.id),
  )

  return {
    initialUsers,
    initialQuestions,
    addedUsers: fetchAddedUsers,
    addedQuestions: fetchAddedQuestions,
  }
}

export const { initialUsers, initialQuestions, addedUsers, addedQuestions } =
  await setupTestData()
