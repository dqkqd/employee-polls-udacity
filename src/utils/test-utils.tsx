import type { PreloadedState } from "@reduxjs/toolkit"
import { configureStore } from "@reduxjs/toolkit"
import { render, type RenderOptions } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import React from "react"
import { Provider } from "react-redux"
import { createMemoryRouter, RouterProvider } from "react-router-dom"
import type { AppStore, RootState } from "../app/store"
import authSlice from "../features/auth/authSlice"
import questionReducer from "../features/questions/questionsSlice"
import usersSlice from "../features/users/usersSlice"
import { routesConfig } from "../routes"
import { initialUsers } from "./test-data"

// As a basic setup, import your same slice reducers
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  route?: string
  router?: ReturnType<typeof createMemoryRouter>
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    route = "/",
    preloadedState = {
      users: initialUsers,
    },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        auth: authSlice,
        users: usersSlice,
        questions: questionReducer,
      },
      preloadedState,
    }),
    router = createMemoryRouter(routesConfig, {
      initialEntries: [route],
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper(): JSX.Element {
    return (
      <Provider store={store}>
        <RouterProvider router={router} />
      </Provider>
    )
  }

  window.history.pushState({}, "Test page", route)

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    user: userEvent.setup(),
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  }
}

export function renderDefault({
  route = "/",
  preloadedState = {
    users: initialUsers,
  },
  // Automatically create a store instance if no store was passed in
  store = configureStore({
    reducer: {
      auth: authSlice,
      users: usersSlice,
      questions: questionReducer,
    },
    preloadedState,
  }),
  router = createMemoryRouter(routesConfig, {
    initialEntries: [route],
  }),
  ...renderOptions
}: ExtendedRenderOptions = {}) {
  return renderWithProviders(<></>, {
    route,
    preloadedState,
    store,
    router,
    ...renderOptions,
  })
}
