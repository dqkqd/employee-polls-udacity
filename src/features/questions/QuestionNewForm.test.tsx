import { act, screen, waitFor } from "@testing-library/react"
import { describe, expect, it } from "vitest"
import { getQuestions, getUser } from "../../api"
import { Question, User } from "../../interfaces"
import {
  initialAuth,
  initialQuestions,
  initialUsers,
} from "../../utils/test-data"
import { renderDefault, renderWithNoRoutes } from "../../utils/test-utils"
import QuestionNewForm from "./QuestionNewForm"

it("Render", () => {
  const { baseElement } = renderWithNoRoutes(<QuestionNewForm />)
  expect(baseElement).toMatchSnapshot()

  expect(
    screen.getByRole("textbox", { name: "First option" }),
  ).toBeInTheDocument()
  expect(
    screen.getByRole("textbox", { name: "Second option" }),
  ).toBeInTheDocument()
  expect(screen.getByRole("button", { name: "Submit" })).toBeInTheDocument()
})

describe("Input", () => {
  it("Write to form should change text", async () => {
    const { user } = renderWithNoRoutes(<QuestionNewForm />)

    const firstOptionEle = screen.getByRole("textbox", { name: "First option" })

    expect(firstOptionEle).toHaveDisplayValue("")
    await act(async () => {
      await user.type(firstOptionEle, "@new-question-option-1")
    })
    expect(firstOptionEle).toHaveDisplayValue("@new-question-option-1")

    const secondOptionEle = screen.getByRole("textbox", {
      name: "Second option",
    })
    expect(secondOptionEle).toHaveDisplayValue("")
    await act(async () => {
      await user.type(secondOptionEle, "@new-question-option-2")
    })
    expect(secondOptionEle).toHaveDisplayValue("@new-question-option-2")
  })
})

describe("Button", () => {
  it("Should be disabled when one of the input is empty", async () => {
    const { user } = renderWithNoRoutes(<QuestionNewForm />)
    const button = screen.getByRole("button", { name: "Submit" })
    expect(button).toBeDisabled()

    const firstOptionEle = screen.getByRole("textbox", { name: "First option" })
    const secondOptionEle = screen.getByRole("textbox", {
      name: "Second option",
    })

    await act(async () => {
      await user.type(firstOptionEle, "123")
      await user.clear(secondOptionEle)
    })
    expect(button).toBeDisabled()

    await act(async () => {
      await user.clear(firstOptionEle)
      await user.type(secondOptionEle, "123")
    })
    expect(button).toBeDisabled()

    await act(async () => {
      await user.type(firstOptionEle, "123")
      await user.type(secondOptionEle, "1234")
    })
    expect(button).toBeEnabled()
  })

  it("Should be disabled when options are the same", async () => {
    const { user } = renderWithNoRoutes(<QuestionNewForm />)
    const button = screen.getByRole("button", { name: "Submit" })

    const firstOptionEle = screen.getByRole("textbox", { name: "First option" })
    const secondOptionEle = screen.getByRole("textbox", {
      name: "Second option",
    })

    await act(async () => {
      await user.type(firstOptionEle, "123")
      await user.type(secondOptionEle, "123")
    })
    expect(button).toBeDisabled()
  })
})

describe("Error", () => {
  it("Should show error when two options are the same", async () => {
    const { user } = renderWithNoRoutes(<QuestionNewForm />)
    expect(
      screen.queryByText("Two options should be different"),
    ).not.toBeInTheDocument()

    const firstOptionEle = screen.getByRole("textbox", { name: "First option" })
    const secondOptionEle = screen.getByRole("textbox", {
      name: "Second option",
    })

    await act(async () => {
      await user.type(firstOptionEle, "123")
      await user.type(secondOptionEle, "123")
    })
    expect(
      screen.getByText("Two options should be different"),
    ).toBeInTheDocument()

    // with space
    await act(async () => {
      await user.clear(firstOptionEle)
      await user.type(firstOptionEle, "   123  ")
    })
    expect(
      screen.getByText("Two options should be different"),
    ).toBeInTheDocument()

    await act(async () => {
      await user.clear(firstOptionEle)
      await user.type(firstOptionEle, "1234")
    })
    expect(
      screen.queryByText("Two options should be different"),
    ).not.toBeInTheDocument()
  })
})

describe("Submit", () => {
  it("Successfully submit question", async () => {
    const auth = initialAuth

    const { store, user } = renderDefault({
      route: "/add",
      preloadedState: {
        users: initialUsers,
        questions: initialQuestions,
        auth,
      },
    })

    const firstOptionEle = screen.getByRole("textbox", { name: "First option" })
    const secondOptionEle = screen.getByRole("textbox", {
      name: "Second option",
    })
    const button = screen.getByRole("button", { name: "Submit" })

    await act(async () => {
      await user.type(firstOptionEle, "123")
      await user.type(secondOptionEle, "1234")
      await user.click(button)
    })

    // submitting up user, everything should be disabled
    expect(screen.getByTestId("add-question-loading")).toBeInTheDocument()
    expect(firstOptionEle).toBeDisabled()
    expect(secondOptionEle).toBeDisabled()
    expect(button).toBeDisabled()

    // should moved into question detail page
    await waitFor(() => {
      expect(screen.getByText(/poll by you/i)).toBeInTheDocument()
    })

    // question should be saved in the store
    const { questions, users } = store.getState()
    // new questions should be on top
    const questionId = questions.ids[0]
    const question = questions.entities[questionId] as Question

    // it should be created in less than 10 seconds ago
    const now = new Date().getTime()
    expect(now - question.timestamp).toBeLessThan(10000)

    const employee = users.entities[auth.id as string] as User
    expect(employee.questions).toContain(questionId)

    // new questions should be in store
    await act(async () => {
      const userFromDatabase = await getUser(auth.id as string)
      expect(userFromDatabase.questions).toContain(questionId)

      const questionFromDatabase = await getQuestions()
      expect(questionFromDatabase[questionId]).toBeDefined()
    })
  })
})
