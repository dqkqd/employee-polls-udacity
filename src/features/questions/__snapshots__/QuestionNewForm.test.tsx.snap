// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render 1`] = `
<body>
  <div>
    <div
      class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          First option
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                First option
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r1:"
          id=":r1:-label"
        >
          Second option
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r1:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Second option
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled=""
      tabindex="-1"
      type="submit"
    >
      Submit
    </button>
  </div>
</body>
`;
